bin/docker.js

```diff
@@ -0,0 +1,95 @@
+#! /usr/bin/env node
+/* eslint-disable no-console */
+const util = require('util'); // eslint-disable-line unicorn/import-style
+const exec = util.promisify(require('child_process').exec);
+
+/**
+ * Retrieves and parses the docker image metadata
+ * @see {@link https://github.com/docker/metadata-action}
+ */
+function getMetadata() {
+  try {
+    const raw = process.env.DOCKER_METADATA_OUTPUT_JSON;
+    const metadata = JSON.parse(raw);
+    if (!Object.keys(metadata.labels || {})?.length) {
+      throw new Error('Invalid shape (missing labels data)');
+    }
+    if (!metadata?.tags?.length) {
+      throw new Error('Invalid shape (missing tags data)');
+    }
+    return metadata;
+  } catch (e) {
+    console.error('Error retrieving docker metadata:', e.message);
+    return process.exit(1);
+  }
+}
+
+/**
+ * Runs command and logs all output
+ */
+async function runCmd(cmd) {
+  // https://stackoverflow.com/a/63027900
+  const execCmd = exec(cmd);
```

Comment:
`child_process.exec()` takes a string that it passes to the shell. However that's not particularly safe, and we're not doing anything to sanitize the input when building the shell commands in `run()`.

> Never pass unsanitized user input to this function. Any input containing shell metacharacters may be used to trigger arbitrary command execution.

Source: https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback

Can we switch these commands to `child_process.execFile` and pass an array of arguments? https://nodejs.org/api/child_process.html#child_processexecfilefile-args-options-callback

Reply:
So I might be misunderstanding but I'm not sure arbitrary command execution is much of an issue here — I already do some sanitization checks here to ensure that the env variable input is valid (and this is really the only "user" input): https://github.com/readmeio/rdme/blob/52d80d4640f2900c474c345dda4382cdee0eb40c/bin/docker.js#L16-L21

And beyond that, everything is mostly hardcoded strings to construct the docker CLI commands and call the `runDockerCmd` (f.k.a. `runCmd`) function (and I've ran through the commands locally a bunch to confirm that they're valid).

In any case, I refactored this in https://github.com/readmeio/rdme/pull/777/commits/1b07bb7def65224317d2dc1f38e88516a46af38f to use `execFile`, TIL! It's **much** more readable now too so thanks for the suggestion 🎊

Reply:
Looks good! My thinking was the tags, which come from git branch names and those could be anything. It looks like git does have some sanitization (I tried a semicolon and got `fatal: 'foo; reboot' is not a valid branch name`) but I could imagine there are ways around this if someone was motivated to try hard enough.